From 7b505fc9e0e6f098ca788236b313247b1d3d0735 Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Wed, 18 Sep 2024 18:04:33 +0200
Subject: [PATCH 1240/1695] drivers: stm32_rng: use default PRNG init if RNG is
 disabled in the DT

In case RNG peripheral is not enabled in device tree, fallback to the
default RNG initialization implementation.
This is useful for M33TD use cases since TFM is not providing a random
number to seed the software RNG of OP-TEE.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 98c1ce52733ba2774b26ee3636f052fa967c7e17 drivers: stm32_rng: use default PRNG init if RNG is disabled in the DT]
Change-Id: I42eb26e1c68ac0daf2adb19b501f681fefa3aeda
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/405415
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/stm32_rng.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index c557159f7..468e810f2 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -467,11 +467,17 @@ out:
 }
 
 #ifdef CFG_WITH_SOFTWARE_PRNG
-/* Override weak plat_rng_init with platform handler to seed PRNG */
+/* Override weak plat_rng_init with platform handler to attempt to seed PRNG */
 void plat_rng_init(void)
 {
 	uint8_t seed[RNG_FIFO_BYTE_DEPTH] = { };
 
+	if (!stm32_rng) {
+		__plat_rng_init();
+		DMSG("PRNG seeded without RNG");
+		return;
+	}
+
 	if (stm32_rng_read(seed, sizeof(seed)))
 		panic();
 
-- 
2.47.2

